### PR TITLE
Crusade Against Stupid Switch Statements: Part 0 of N

### PR DESCRIPTION
**Pull request by VictorSohier**
Found a stupid switch statement with some stupid templates and thought why not make this N^2 problem a 2N problem. It is currently at the point where it is a little more than a superset of the original switch statement. There are some placeholders for additional conversion routines and SIMD options with fallbacks to scalar implementations. This also contains some additional generic machinery for future IO tasks.

## Summary by CodeRabbit

* **New Features**
  * Added buffered image reading and writing for supported uncompressed formats.
  * Improved image format conversion across all mipmap levels.
  * Added support for reading and writing image data from memory-backed sources.
  * Added safer handling for partial image blocks, unsupported formats, and I/O failures.

* **Tests**
  * Expanded coverage for memory-slice operations, overlapping copies, typed access, and boundary conditions.

### CODE REVIEW

File: core/io/image_rw.cpp [hunk 1]
```diff
@@ -0,0 +1,1031 @@
...
+		uint32_t height) noexcept {
+	dst->reader = r;
+	if (__builtin_cpu_supports("avx2")) {
+		dst->vtbl = &(READER_AVX2_FUNCTIONS[format]);
+	} else if (__builtin_cpu_supports("avx")) {
+		dst->vtbl = &(READER_AVX_FUNCTIONS[format]);
+	} else if (__builtin_cpu_supports("sse4.2")) {
+		dst->vtbl = &(READER_SSE42_FUNCTIONS[format]);
+	} else {
+		dst->vtbl = &(READER_SCALAR_FUNCTIONS[format]);
```
[the remark this answers is not in the document]
Comment (VictorSohier) [reply core/io/image_rw.cpp:1135]: Give me the correct compiler guards.

File: core/io/image_rw.cpp [hunk 2]
```diff
@@ -0,0 +1,1031 @@
...
+	}
+	if (dst->vtbl->read == nullptr) {
+		return IO::Error::NotImplemented;
+	}
+	if (READER_STATE_CONSTRUCTOR[format]) {
+		return READER_STATE_CONSTRUCTOR[format](&(dst->state), width, height, format);
+	}
+	return IO::Error::Okay;
```
[the remark this answers is not in the document]
Comment (VictorSohier) [reply core/io/image_rw.cpp:1146]: The staging buffer is, in fact, uninitialized. This is correct behavior for the current readers.

File: core/io/image_rw.cpp [hunk 3]
```diff
@@ -0,0 +1,1031 @@
...
+		for (size_t j = 0; j < 4; j += 1) {
+			if ((((block << 2) + (width * i)) + j) < (width * (i + 1))) {
+				(*sliceAt(
+						data,
+						uint8_t,
+						(((block << 2) + (width * i)) + j))) = (13938U * ((uint64_t)colors->r[(i << 2) + j] * 255) +
+																	   46869U * ((uint64_t)colors->g[(i << 2) + j] * 255) +
+																	   4729U * ((uint64_t)colors->b[(i << 2) + j] * 255) +
+																	   32768U) >>
+						16U;
```
[the remark this answers is not in the document]
Comment (VictorSohier) [reply core/io/image_rw.cpp]: Is this in the reader or writer you are talking about and does it still exist in the current revision?

File: core/io/image_rw.cpp [hunk 4]
```diff
@@ -0,0 +1,1175 @@
...
+	row = state->currentRow;
+	for (; i.i < 16; i.i += 1) {
+		if (i.col == 0) {
+			b = 0;
+		}
+		if ((((block << 2) + i.col) < width) & ((row + i.row) < height)) {
+			(*sliceAt(data, uint8_t, (i.row * width) + (block << 2) + b)) = (uint8_t)(((13938U * ((uint64_t)(colors->r[i.p] * 255.0))) +
+																							  (46869U * ((uint64_t)(colors->g[i.p] * 255.0))) +
+																							  (4729U * ((uint64_t)(colors->b[i.p] * 255.0))) + 32768U) >>
+					16);
```
[the remark this answers is not in the document]
Comment (VictorSohier) [reply core/io/image_rw.cpp]: This PR does not set out to fix existing bugs, merely to make the existing code more maintainable. If this ever becomes a problem, it will be fixed later. There is even a assertion in the tests for the one edge case where this may become a problem and it passes.

[the remark this answers is not in the document]
Comment (VictorSohier) [reply core/io/image_rw.cpp]: @Arctis-Fireblight

File: core/templates/Slice.h [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+inline bool Slice::set(Slice dst, Slice src, size_t index) noexcept {
+	Slice tmp = {};
+	if (!subslice(&tmp, dst, index, src.length)) {
+		return false;
+	}
+	Slice::copy(tmp, src);
+	return true;
+}
+
+#define sliceAt(slice, t, i) ((t *)Slice::get(slice, i, sizeof(t)))
```
Comment (OldDev78) [inline core/templates/Slice.h]: Why can´t these methods above just be be templates on element type instead of relying on macros like this?

```
template <typename T>
constexpr static void *get(
			Slice self, size_t index) noexcept
{
	void *ret = nullptr;
	if ((sizeof(T) * (index + 1)) <= self.length) {
		ret = static_cast<uint8_t*>(self.data) + (sizeof(T) * index);
	}
	return ret;
}
```

You can keep it based on void*

Comment (VictorSohier) [reply core/templates/Slice.h]: I'd rather just have such a function wrap the void * version. I want to have the function be able to be compiled independently of the template. The important bit is that the internals are type erased.

**Comment by Arctis-Fireblight** [comment]
I am a little hesitant to merge a change like this in for 26.3 since there isn't much time left to do through regression testing. May want to hold off until 27.1

File: core/io/image_rw.cpp [hunk 6]
```diff
@@ -0,0 +1,1176 @@
...
+	{ .read = (ReadProc)rgba4444ReadScalar, .destroy = (IO::DestroyProc)Memory::free_aligned_static }, // FORMAT_RGBA4444
+	{ .read = nullptr, .destroy = (IO::DestroyProc)Memory::free_aligned_static }, // FORMAT_RGB565
+	{ .read = (ReadProc)rgbafReadScalar, .destroy = (IO::DestroyProc)Memory::free_aligned_static }, // FORMAT_RF
+	{ .read = (ReadProc)rgbafReadScalar, .destroy = (IO::DestroyProc)Memory::free_aligned_static }, // FORMAT_RGF
+	{ .read = (ReadProc)rgbafReadScalar, .destroy = (IO::DestroyProc)Memory::free_aligned_static }, // FORMAT_RGBF
+	{ .read = (ReadProc)rgbafReadScalar, .destroy = (IO::DestroyProc)Memory::free_aligned_static }, // FORMAT_RGBAF
+	{ .read = (ReadProc)rgbahReadScalar, .destroy = (IO::DestroyProc)Memory::free_aligned_static }, // FORMAT_RH
+	{ .read = (ReadProc)rgbahReadScalar, .destroy = (IO::DestroyProc)Memory::free_aligned_static }, // FORMAT_RGH
+	{ .read = (ReadProc)rgbahReadScalar, .destroy = (IO::DestroyProc)Memory::free_aligned_static }, // FORMAT_RGBH
+	{ .read = (ReadProc)rgbahReadScalar, .destroy = (IO::DestroyProc)Memory::free_aligned_static }, // FORMAT_RGBAH
```
[the remark this answers is not in the document]
Comment (VictorSohier) [reply core/io/image_rw.cpp:709]: This is FINE, this is a form of type reification, it just happens on call as opposed to needing an explicit cast. Also, these routines do work as there are end to end image conversion tests that pass.

File: core/io/image_rw.cpp [hunk 7]
```diff
@@ -0,0 +1,1176 @@
...
+			struct
+			{
+				uint8_t col : 2;
+				uint8_t row : 2;
+				uint8_t c : 2;
+			};
+		};
+		uint8_t _p;
+	};
+};
```
[the remark this answers is not in the document]
Comment (VictorSohier) [reply core/io/image_rw.cpp:63]: No, this is fine, this is much more convenient than having to do all the extra math manually. This code is written to the de facto standard, not the de jure standard.

File: core/io/image_rw.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (VictorSohier) [reply core/io/image_rw.cpp:709]: There are only so many things that can happen to a pointer across a function boundary. It isn't like the value of the pointer changes just because it goes from `void *` to `UncompressedImageState *`. Also, are you sure that it will always be the same state, because I'm not sure and I'm the one who is performing this conversion. I call this specific criticism invalid.

File: core/io/image_rw.cpp [hunk 8]
```diff
@@ -0,0 +1,1176 @@
...
+}
+
+static IO::Error defaultFlush(
+		IO::Writer writer,
+		UncompressedImageState *state) noexcept {
+	IO::Error err = IO::Error::Okay;
+	size_t cursor;
+	size_t row = state->currentRow;
+	Slice tmp = {};
+	err = IO::Writer::seek(writer, 0, &cursor, IO::WHENCE_CURRENT);
```
[the remark this answers is not in the document]
Comment (VictorSohier) [reply core/io/image_rw.cpp]: Resolved after reading it over again.

File: core/templates/slice.h [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+	void *ret = nullptr;
+	if ((size * (index + 1)) <= self.length) {
+		ret = ((uint8_t *)self.data) + (size * index);
+	}
+	return ret;
+}
+
+template <typename T>
+constexpr inline T *Slice::at(Slice self, size_t index) noexcept {
+	return (T *)Slice::get(self, index, sizeof(T));
```
[the remark this answers is not in the document]
Comment (VictorSohier) [reply core/templates/slice.h:88]: Honestly, I don't know how often this error condition will be hit, especially considering that this is a structure using 64 bit values in a 48 bit address space (point to an architecture that does otherwise and I'll eat my words). IIRC, we don't support 32 bit targets right now. Yes, I do know about the classic binary search overflow bug. I'm pretty sure that in order to run into this, you have other problems elsewhere.

Also, I'd much prefer it that, if you caught this in the initial code review that you drop everything in the initial code review. I have seen this multiple times before where there are bugs that persist between code rabbit code reviews that only get brought up later.

[the remark this answers is not in the document]
Comment (VictorSohier) [reply core/templates/slice.h:88]: Thinking about it, the fix is likely less safe is you'd then be more likely to underflow the **unsigned** values though subtraction. I'll leave it as is for this reason.
